CASSSIDECAR-329: Sidecar endpoint for vending statistics related to compaction

sklaha opened the pull request:
Sidecar endpoint for vending statistics related to compaction

These changes are part of the effort to introduce bespoke Sidecar APIs to support key operational functionality currently managed through nodetool commands. Changes in the commit:
* API for vending compcation stats
* Unit tests
* Integration tests

Circle CI build: https://app.circleci.com/pipelines/github/sklaha/cassandra-sidecar/10/workflows/032ca3cf-e90d-413a-9f77-beb01f696be6

Commit:
CASSSIDECAR-329: Sidecar endpoint for vending statistics related to compaction
These changes are part of the effort to introduce bespoke Sidecar APIs to support key operational functionality currently managed through nodetool commands.
Changes in the commit:
* API for vending compcation stats
* Unit tests
* Integration tests

frankgh commented on server/src/main/java/org/apache/cassandra/sidecar/handlers/CompactionStatsHandler.java:
```diff
@@ -0,0 +1,82 @@
+                     .onFailure(cause -> processFailure(cause, context, host, remoteAddress, request));
+    }
+
+    protected Void extractParamsOrThrow(RoutingContext context)
```
include override annotation
```suggestion
    @Override
    protected Void extractParamsOrThrow(RoutingContext context)
```

sklaha replied:
Done

yifan-c commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/jmx/StorageJmxOperations.java:
```diff
@@ -198,4 +198,37 @@ public interface StorageJmxOperations
+     * Returns the compaction throughput limit in MiB per second as a double
+     * @return compaction throughput limit
+     */
+    double getCompactionThroughputMbPerSec();
```
not used. Please explain if it is needed

sklaha replied:
I removed it.

yifan-c commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/jmx/StorageJmxOperations.java:
```diff
@@ -198,4 +198,37 @@ public interface StorageJmxOperations
+     * Returns the compaction throughput limit in MiB per second as a double (alternative method name)
+     * @return compaction throughput limit
+     */
+    double getCompactionThroughtputMibPerSecAsDouble();
```
not used. Please explain if it is needed

sklaha replied:
I removed it.

yifan-c commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/jmx/StorageJmxOperations.java:
```diff
@@ -198,4 +198,37 @@ public interface StorageJmxOperations
+     * Returns current compaction throughput measurements in MiB per second for different time windows
+     * @return map of time windows to throughput measurements
+     */
+    Map<String, String> getCurrentCompactionThroughputMebibytesPerSec();
```
not used. Please explain if it is needed

sklaha replied:
I removed it.

yifan-c commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/jmx/StorageJmxOperations.java:
```diff
@@ -198,4 +198,37 @@ public interface StorageJmxOperations
+     * 
+     * @return the current compaction throughput in bytes per second, or 0 if throughput cannot be determined
+     */
+    long getCompactionThroughtputBytesPerSec();
```
Please add a comment why there is a typo. Otherwise, it might be corrected by mistake in the future.

sklaha replied:
Added a comment on that.

yifan-c commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/jmx/CompactionManagerJmxOperations.java:
```diff
@@ -0,0 +1,48 @@
+     * @return compaction throughput in bytes per second
+     */
+    long getCompactionThroughputMbPerSec();
+}
```
new line at the end of file per sytle guide

sklaha replied:
I fixed all checkstyle issues.

yifan-c commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/data/CompositeDataUtil.java:
```diff
@@ -50,4 +50,55 @@ public static <T> T extractValue(CompositeData data, String key)
+    {
+        if (!expectedType.isInstance(value))
+        {
+            throw new IllegalStateException("Expected " + expectedType.getSimpleName() + " for " + contextDescription + " but got: " +
+                    (value == null ? "null" : value.getClass().getSimpleName()));
```
Should it throw `ClassCastException` instead? Looks like the type check here is to prepare a better description of the error.

sklaha replied:
done

yifan-c commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/data/CompositeDataUtil.java:
```diff
@@ -50,4 +50,55 @@ public static <T> T extractValue(CompositeData data, String key)
+     * @throws IllegalStateException if the value cannot be parsed as a long,
+     *         with a descriptive message indicating what failed to parse
+     */
+    public static long safeParseLong(final String value, final String contextDescription)
```
remove `final`

sklaha replied:
done

yifan-c commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/data/CompositeDataUtil.java:
```diff
@@ -50,4 +50,55 @@ public static <T> T extractValue(CompositeData data, String key)
+        }
+        catch (NumberFormatException e)
+        {
+            throw new IllegalStateException("Failed to parse long value '" + value + "' for " + contextDescription + ": " + e.getMessage(), e);
```
re-throw `NumberFormatException` with better description?

sklaha replied:
done

yifan-c commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/data/CompositeDataUtil.java:
```diff
@@ -50,4 +50,55 @@ public static <T> T extractValue(CompositeData data, String key)
+            throw new IllegalStateException("Failed to parse long value '" + value + "' for " + contextDescription + ": " + e.getMessage(), e);
+        }
+    }
+
```
nit: remove empty line.

sklaha replied:
Done

yifan-c commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -61,6 +67,27 @@ public class CassandraMetricsOperations implements MetricsOperations
+    public static final String TIME_FORMAT = "%dh%02dm%02ds";
+
+    // Default values
+    public static final String DEFAULTVAL_STRING = "unknown";
+    public static final String DEFAULTVAL_NUMBER = "-1";
+    public static final String DEFAULTVAL_N_A = "n/a";
```
`DEFAULT_STRING_VALUE`? and so on.

sklaha replied:
Done

yifan-c commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -250,4 +278,187 @@ private List<ClientConnectionEntry> statsToEntries(Stream<ConnectedClientStats>
+        // Calculate remaining time in seconds based on throughput and remaining bytes
+        String activeCompactionsRemainingTime = calculateRemainingTimeSeconds(activeCompactions, storageService);
+
+        return new CompactionStatsResponse(concurrentCompactors, pendingTasks, totalPendingTasks,
+                                           completedCompactions, dataCompacted, abortedCompactions,
+                                           reducedCompactions, sstablesDroppedFromCompaction,
+                                           completedCompactionsRate,
+                                           activeCompactions, activeCompactionsCount, activeCompactionsRemainingTime);
```
strongly -1 on returning Http response object from this layer (jmx). It should return a data structure (say, `CompactionStats`) can be shared by other server components. 
The handlers are to return the http response object (converting from `CompactionStats`)

sklaha replied:
I agree that we could utilize data access objects in the JMX layer. However, I opted to return the response object for the following reasons:

- Other APIs are doing the same.
- Incorporating data access objects would necessitate the creation of object mappers.

We now have two options to consider:

1. Maintain the current practice of returning the response object from JMX. When we require reusing the object for some other internal purpose, we can then create the data access objects and mappers.
2. Create the data access objects and mappers now.

Please share your thoughts on this matter.

frankgh replied:
I agree with Yifan here. We shouldn't leak the http layer to the adapter layer. Also, can you leverage the builder pattern here by leveraging `org.apache.cassandra.sidecar.common.DataObjectBuilder`

sklaha replied:
Done

yifan-c commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -250,4 +278,187 @@ private List<ClientConnectionEntry> statsToEntries(Stream<ConnectedClientStats>
+        String metricObjectType = String.format(METRICS_OBJ_TYPE_COMPACTION, "TotalCompactionsCompleted");
+        MeterMetricsJmxOperations metricsProxy = jmxClient.proxy(MeterMetricsJmxOperations.class, metricObjectType);
+
+        // Convert rates according to specification:
```
```suggestion
        // Convert per-second rates to the specification:
```

sklaha replied:
Done

yifan-c commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -250,4 +278,187 @@ private List<ClientConnectionEntry> statsToEntries(Stream<ConnectedClientStats>
+    }
+
+
+    private List<ActiveCompactionEntry> getActiveCompactions(final List<Map<String, String>> compactions)
```
no `final` in the parameter

sklaha replied:
Done

yifan-c commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -250,4 +278,187 @@ private List<ClientConnectionEntry> statsToEntries(Stream<ConnectedClientStats>
+            // Parse SSTables list
+            String ssTablesStr = compactionInfo.getOrDefault(SSTABLES, DEFAULTVAL_STRING);
+            List<String> ssTables = ssTablesStr.isEmpty() ?
+                List.of() : List.of(ssTablesStr.split(","));
```
```suggestion
            List<String> ssTables = ssTablesStr.isEmpty() ? List.of() : List.of(ssTablesStr.split(","));
```

sklaha replied:
Done

yifan-c commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -250,4 +278,187 @@ private List<ClientConnectionEntry> statsToEntries(Stream<ConnectedClientStats>
+            double percentCompleted = totalBytes == -1 ? 0.0 : (double) completedBytes / totalBytes * 100.0;
+            
+            // Parse SSTables list
+            String ssTablesStr = compactionInfo.getOrDefault(SSTABLES, DEFAULTVAL_STRING);
```
Does it return `List("unknown")` when there is no sstable? Should it return empty list instead?

sklaha replied:
Yes. Actually, making all string values blank will be better to avoid confusion. Updated it.

yifan-c commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -250,4 +278,187 @@ private List<ClientConnectionEntry> statsToEntries(Stream<ConnectedClientStats>
+        }
+
+        // Calculate time in seconds (throughput is already in bytes/sec)
+        long remainingTimeInSecs = totalRemainingBytes / throughputBytesPerSec;
```
Convert to double for the time calculation? otherwise, the result is off by 1.

sklaha replied:
done

yifan-c commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/request/CompactionStatsRequest.java:
```diff
@@ -0,0 +1,46 @@
+    {
+        return HttpMethod.GET;
+    }
+}
```
new line at the EOF

sklaha replied:
Done

yifan-c commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/response/CompactionStatsResponse.java:
```diff
@@ -0,0 +1,194 @@
+    {
+        return activeCompactionsRemainingTime;
+    }
+}
```
new line

sklaha replied:
Done

yifan-c commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/response/data/ActiveCompactionEntry.java:
```diff
@@ -0,0 +1,131 @@
+     * @param targetDirectory  target directory for output
+     */
+    @JsonCreator
+    public ActiveCompactionEntry(@JsonProperty("id") final String id,
+                                 @JsonProperty("keyspace") final String keyspace,
+                                 @JsonProperty("columnFamily") final String columnFamily,
+                                 @JsonProperty("taskType") final String taskType,
+                                 @JsonProperty("completedBytes") final long completedBytes,
+                                 @JsonProperty("totalBytes") final long totalBytes,
+                                 @JsonProperty("percentCompleted") final double percentCompleted,
+                                 @JsonProperty("ssTables") final List<String> ssTables,
+                                 @JsonProperty("targetDirectory") final String targetDirectory)
```
no `final`

sklaha replied:
Done

yifan-c commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/response/data/ActiveCompactionEntry.java:
```diff
@@ -0,0 +1,131 @@
+    {
+        return targetDirectory;
+    }
+}
```
new line

sklaha replied:
I fixed all checkstyle issues.

yifan-c commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/response/data/PendingCompactionTasks.java:
```diff
@@ -0,0 +1,61 @@
+    {
+        return totalPendingTasks;
+    }
+}
```
new line

sklaha replied:
I fixed all checkstyle issues.

yifan-c commented on integration-tests/src/integrationTest/org/apache/cassandra/sidecar/routes/CompactionStatsIntegrationTest.java:
```diff
@@ -0,0 +1,337 @@
+    {
+        createTestKeyspace(TEST_KEYSPACE, DC1_RF1);
+
+        for (int i = 1; i <= TABLE_COUNT; i++) {
+            TEST_TABLES.add(new QualifiedName(TEST_KEYSPACE, TEST_TABLE_PREFIX + "_compaction_" + i));
+        }
+        
+        // Create test tables for compaction activity
+        for (QualifiedName tableName : TEST_TABLES) {
+            createTestTable(tableName,
+                            "CREATE TABLE %s ( \n" +
+                            "  id int PRIMARY KEY, \n" +
+                            "  data text \n" +
+                            ");");
+        }
+    }
```
left curly brace at new line (this and others in this file)

sklaha replied:
I fixed all checkstyle issues.

yifan-c commented on server/src/main/java/org/apache/cassandra/sidecar/handlers/CompactionStatsHandler.java:
```diff
@@ -0,0 +1,82 @@
+    {
+        return null;
+    }
+}
```
new line

sklaha replied:
I fixed all checkstyle issues.

yifan-c commented on server/src/main/java/org/apache/cassandra/sidecar/modules/CassandraOperationsModule.java:
```diff
@@ -90,6 +92,22 @@ VertxRoute cassandraConnectedClientStatsRoute(RouteBuilder.Factory factory,
         return factory.buildRouteWithHandler(connectedClientStatsHandler);
     }
 
+    @GET
+    @Path(ApiEndpointsV1.COMPACTION_STATS_ROUTE)
+    @Operation(summary = "Get compaction statistics",
+               description = "Returns compaction statistics for the Cassandra node")
+    @APIResponse(description = "Compaction statistics retrieved successfully",
+                 responseCode = "200",
+                 content = @Content(mediaType = "application/json",
+                 schema = @Schema(implementation = CompactionStatsResponse.class)))
```
👍

Commit:
Fixed checkstyle errors

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -250,4 +295,186 @@ private List<ClientConnectionEntry> statsToEntries(Stream<ConnectedClientStats>
+    /**
+     * Represents the metrics related to compaction stats that are supported by the Sidecar
+     */
+    public enum CompactionStatsMetrics
```
this feels misplaced, move it to its own class file

sklaha replied:
Done

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -250,4 +295,186 @@ private List<ClientConnectionEntry> statsToEntries(Stream<ConnectedClientStats>
+    public CompactionStatsResponse compactionStats()
+    {
+        // Get compaction manager and storage service proxies
+        CompactionManagerJmxOperations compactionManager = jmxClient.proxy(CompactionManagerJmxOperations.class, COMPACTION_MANAGER_OBJ_NAME);
```
we should segregate the logic to get the compaction stats from this class. This class should be a thin layer that proxies the request into another class where you have the business logic to retrieve compaction stats

sklaha replied:
Would you like to have a separate class which will have the business logic? Or would like like me to keep the business logic here and move the data retrieval to somewhere else? Please note that we compaction stats require multiple JMX objects like storage, compaction, metrics for the data.

frankgh replied:
yes, definitely. We should follow the single responsibility principle here. The class is becoming overloaded

sklaha replied:
Done.

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -250,4 +295,186 @@ private List<ClientConnectionEntry> statsToEntries(Stream<ConnectedClientStats>
+
+        // Get compaction metrics from JMX counters and meters
+        long completedCompactions = getValueAsLong(getCompactionMetric(CompactionStatsMetrics.TOTAL_COMPACTIONS_COMPLETED));
+        long dataCompacted = getValueAsLong(getCompactionMetric(CompactionStatsMetrics.BYTES_COMPACTED));
```
let's preserve the units in the variable name

sklaha replied:
done

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -250,4 +295,186 @@ private List<ClientConnectionEntry> statsToEntries(Stream<ConnectedClientStats>
+        return parsePendingTasksMap(value);
+    }
+
+    private Map<String, Map<String, Integer>> parsePendingTasksMap(final Object value)
```
NIT
```suggestion
    private Map<String, Map<String, Integer>> parsePendingTasksMap(Object value)
```

sklaha replied:
Done

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
generally seeing a lot of empty lines with whitespace, it would be good to format them

sklaha replied:
I left some empty lines for better readability. Are you referring to those?

frankgh replied:
no no, just the whitespace in empty lines.

sklaha replied:
Fixed

frankgh commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/response/data/ActiveCompactionEntry.java:
```diff
@@ -0,0 +1,131 @@
+{
+    private final String id;
+    private final String keyspace;
+    private final String columnFamily;
```
I think we should use table here instead. columFamily is the deprecated name, but it's probably kept here for backwards compatibility reasons.

sklaha replied:
Done

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -250,4 +295,186 @@ private List<ClientConnectionEntry> statsToEntries(Stream<ConnectedClientStats>
+            long totalBytes = safeParseLong(compactionInfo.getOrDefault(TOTAL, DEFAULT_NUMBER_VALUE), "total bytes");
+            
+            // Calculate percentage completed
+            double percentCompleted = totalBytes == -1 ? 0.0 : (double) completedBytes / totalBytes * 100.0;
```
if totalBytes == 0 then your division will throw

sklaha replied:
Fixed.

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -250,4 +295,186 @@ private List<ClientConnectionEntry> statsToEntries(Stream<ConnectedClientStats>
+    {
+        return compactions.stream().map(compactionInfo -> {
+            // Extract fields according to specification
+            String id = compactionInfo.getOrDefault(COMPACTION_ID, DEFAULT_STRING_VALUE);
```
not sure why we are defaulting to empty string here? should it just be null instead?

sklaha replied:
I don't have a strong preference here. Do you think setting the default null will be better?

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -250,4 +295,186 @@ private List<ClientConnectionEntry> statsToEntries(Stream<ConnectedClientStats>
+        }).collect(Collectors.toList());
+    }
+
+    private String calculateRemainingTimeSeconds(final List<ActiveCompactionEntry> activeCompactions,
+                                               final StorageJmxOperations storageService)
```
NIT remove final keyword
```suggestion
    private String calculateRemainingTimeSeconds(List<ActiveCompactionEntry> activeCompactions,
                                                StorageJmxOperations storageService)
```

sklaha replied:
Done

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -250,4 +295,186 @@ private List<ClientConnectionEntry> statsToEntries(Stream<ConnectedClientStats>
+        long activeCompactionsCount = activeCompactions.size();
+        
+        // Calculate remaining time in seconds based on throughput and remaining bytes
+        String activeCompactionsRemainingTime = calculateRemainingTimeSeconds(activeCompactions, storageService);
```
this should be a long value. And the formatting should happen at the client side. Your raw values should be preserved in the payload you send to the client

sklaha replied:
Done

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/data/CompositeDataUtil.java:
```diff
@@ -50,4 +50,55 @@ public static <T> T extractValue(CompositeData data, String key)
+     * @throws IllegalStateException if the value is not an instance of the expected type,
+     *         with a descriptive message indicating what was expected vs what was received
+     */
+    public static <T> T safeCast(Object value, Class<T> expectedType, String contextDescription)
```
we should have unit tests here. I was trying to understand what happens when you do `expectedType.isInstance(null)`. A unit test should give me an idea of what the behavior will be

sklaha replied:
Added unit tests.

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/jmx/CompactionManagerJmxOperations.java:
```diff
@@ -0,0 +1,48 @@
+    List<Map<String, String>> getCompactions();
+
+    /**
+     * Returns compaction throughput in bytes per second
```
it looks like the javadoc here is incorrect. Also, this method is not used. Do we need it?

sklaha replied:
Removed.

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/jmx/CompactionManagerJmxOperations.java:
```diff
@@ -0,0 +1,48 @@
+     * Returns the number of concurrent compactors configured for the node.
+     * @return number of concurrent compactors
+     */
+    int getCoreCompactorThreads();
```
unused, remove it?

sklaha replied:
Removed.

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/jmx/CounterMetricsJmxOperations.java:
```diff
@@ -21,16 +21,8 @@
  * An interface that pulls methods from the Cassandra Metrics Proxy
  */
-public interface MetricsJmxOperations
+public interface CounterMetricsJmxOperations
```
not sure if I understand the motivation to split this interface into two, but it seems like keeping the original interface is sufficient?

sklaha replied:
Without the separation, there is a risk of someone accidentally calling getCount() on GaugeMetrics or getValue() on CounterMetrics etc.

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/jmx/MeterMetricsJmxOperations.java:
```diff
@@ -0,0 +1,57 @@
+     * Returns the one-minute exponentially-weighted moving average rate.
+     * @return the one-minute rate in events per second
+     */
+    double getOneMinuteRate();
```
unused

sklaha replied:
Removed.

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/jmx/MeterMetricsJmxOperations.java:
```diff
@@ -0,0 +1,57 @@
+     * Returns the five-minute exponentially-weighted moving average rate.
+     * @return the five-minute rate in events per second
+     */
+    double getFiveMinuteRate();
```
unused

sklaha replied:
Removed.

frankgh commented on client/src/main/java/org/apache/cassandra/sidecar/client/SidecarClient.java:
```diff
@@ -734,6 +735,20 @@ public CompletableFuture<ConnectedClientStatsResponse> connectedClientStats(Side
+     * @param instance the instance where the request will be executed
+     * @return a completable future of the compaction stats
+     */
+    public CompletableFuture<CompactionStatsResponse> compactionStats(SidecarInstance instance)
```
can you add a test in `org.apache.cassandra.sidecar.client.SidecarClientTest`

sklaha replied:
done

frankgh commented on integration-tests/src/integrationTest/org/apache/cassandra/sidecar/routes/CompactionStatsIntegrationTest.java:
```diff
@@ -0,0 +1,372 @@
+ *   <li><strong>Dual Success:</strong> Test passes whether active compactions are captured or completed</li>
+ * </ul>
+ */
+class CompactionStatsIntegrationTest extends SharedClusterSidecarIntegrationTestBase
```
we should merge this test class with `org.apache.cassandra.sidecar.routes.CassandraStatsIntegrationTest`

sklaha replied:
done

frankgh reviewed (commented):
Looks good so far, I have a few comments

yifan-c commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/response/CompactionStatsResponse.java:
```diff
@@ -0,0 +1,194 @@
+                                   @JsonProperty("completedCompactionsRate") CompletedCompactionsRate completedCompactionsRate,
+                                   @JsonProperty("activeCompactions") List<ActiveCompactionEntry> activeCompactions,
+                                   @JsonProperty("activeCompactionsCount") long activeCompactionsCount,
+                                   @JsonProperty("activeCompactionsRemainingTime") String activeCompactionsRemainingTime)
```
nit: create constants for the strings, since each of them is referenced in multiple places.

sklaha replied:
Done

yifan-c commented on integration-tests/src/integrationTest/org/apache/cassandra/sidecar/routes/CompactionStatsIntegrationTest.java:
```diff
@@ -0,0 +1,372 @@
+        HttpResponse<Buffer> response;
+        boolean foundActiveCompactions;
+        
+        for (int attempt = 0; attempt < MAX_POLL_ATTEMPTS; attempt++)
```
Please check out `org.apache.cassandra.testing.utils.AssertionUtils#loopAssert(int, int, java.lang.Runnable)`

sklaha replied:
/**
     * Run the assertions in a loop until the first success within the timeout.
     * Otherwise, it fails with the last assertion failure.
     *
     * @param timeoutSeconds timeout
     * @param assertions     assertions
     */
    public static void loopAssert(int timeoutSeconds, int delayMillis, Runnable assertions)

I don't want to fail the test if it's unable to catch an active compaction. But I want to fail the test for other assertions. It will be difficult to use loopAssert unless I write another version of it.

yifan-c replied:
Got it.

yifan-c reviewed (commented):
Forgot to comment the last time... Please update `changes.txt`.

Commit:
Addressed review comments on compaction stats API

arjunashok commented on server-common/src/main/java/org/apache/cassandra/sidecar/common/server/utils/CompactionStatsConverter.java:
```diff
@@ -0,0 +1,114 @@
+    {
+        if (data == null)
+        {
+            return null;
```
Do we want to return null here, or fail-fast with a non-200 response code?

sklaha replied:
Yes, it makes sense to fail the request if the data is null by any chance. I updated the code.

arjunashok commented on server-common/src/main/java/org/apache/cassandra/sidecar/common/server/utils/CompactionStatsConverter.java:
```diff
@@ -0,0 +1,114 @@
+        }
+
+        CompactionStatsResponse.CompletedCompactionsRate responseRate = null;
+        if (data.completedCompactionsRate() != null)
```
Likewise, in which scenario do you foresee these stats being null? If it is valid for them to be null for non-error scenarios, eg when there are no compactions ongoing, we should be more explicit about that in the API response.

In your implementation the current behavior is that we return the metrics that we have available (success w/ partial response). I'm wondering if we should fail-fast instead.

sklaha replied:
In my opinion, providing a partial response will still be valuable rather than failing the request. For example, even if compaction rate is unknown, it will be useful to know if there are any active compactions.

arjunashok commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -250,4 +296,181 @@ private List<ClientConnectionEntry> statsToEntries(Stream<ConnectedClientStats>
+        double meanRateValuePerHour = metricsProxy.getMeanRate() * 3600; // Convert per second to per hour
+        double fifteenMinuteRateValuePerMinute = metricsProxy.getFifteenMinuteRate() * 60; // Convert per second to per minute
+
+        String meanRate = String.format("%.2f/hour", meanRateValuePerHour);
+        String fifteenMinuteRate = String.format("%.2f/minute", fifteenMinuteRateValuePerMinute);
```
Any reason we want to append the units suffix with the values? I'd recommend we keep these as numeric values with the unit-specific details encoded into the metric name. eg. `meanHourlyRate`. That way, clients are not expected to parse the values.

yifan-c replied:
Agree with Arjun. 
In addition, the conversion to per hour or per minute should not be done here. In this layer, it should just return the value as-is. 
If per hour and per minute is part of the API spec, could you do the conversion in the http response, `CompactionStatsResponse`?

sklaha replied:
I agree. It will be better to just return the rate from the API. Users can format the data as per requirement. Updated the code.

frankgh committed:
CASSSIDECAR-336: Fix logging regression in Sidecar after CASSSIDECAR-176 (#253)

Patch by Francisco Guerrero; reviewed by Bernardo Botella, Yifan Cai for CASSSIDECAR-336

yifan-c commented on server-common/src/main/java/org/apache/cassandra/sidecar/common/server/data/CompactionStatsData.java:
```diff
@@ -0,0 +1,296 @@
+ */
+public class CompactionStatsData
+{
+    private final long concurrentCompactors;
+    private final Map<String, Map<String, Integer>> pendingTasks;
+    private final long totalPendingTasks;
+    private final long completedCompactions;
+    private final long dataCompacted;
+    private final long abortedCompactions;
+    private final long reducedCompactions;
+    private final long sstablesDroppedFromCompaction;
+    private final CompletedCompactionsRateData completedCompactionsRate;
+    private final List<ActiveCompactionEntryData> activeCompactions;
+    private final long activeCompactionsCount;
+    private final long activeCompactionsRemainingTime;
```
nit: they can be just `public` and the getters can be removed. Not strong on this one. Feel free to ignore.

Commit:
CASSSIDECAR-329: Addressed review comments

Commit:
Fixed unit test

yifan-c commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/response/data/ActiveCompactionEntry.java:
```diff
@@ -0,0 +1,296 @@
+    public static final String COMPLETED_BYTES = "completedBytes";
+    public static final String TOTAL_BYTES = "totalBytes";
+    public static final String PERCENT_COMPLETED = "percentCompleted";
+    public static final String SS_TABLES = "ssTables";
```
It is less common to break up the term `sstable`

```suggestion
    public static final String SSTABLES = "sstables";
```

sklaha replied:
done

yifan-c commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/response/data/ActiveCompactionEntry.java:
```diff
@@ -0,0 +1,296 @@
+         * @param ssTables the {@code ssTables} to set
+         * @return a reference to this Builder
+         */
+        public Builder ssTables(List<String> ssTables)
```
`sstables`

sklaha replied:
done

yifan-c commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/response/data/ActiveCompactionEntry.java:
```diff
@@ -0,0 +1,296 @@
+ * Represents an active compaction entry
+ */
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class ActiveCompactionEntry
```
It would be nice to name it the same name, `CompactionInfo`, as in Cassandra. Easier to correlate.

sklaha replied:
Done

yifan-c commented on integration-tests/src/integrationTest/org/apache/cassandra/sidecar/routes/CassandraStatsIntegrationTest.java:
```diff
@@ -256,4 +280,284 @@ void assertClientStatsResponse(HttpResponse<Buffer> response, Map<String, Boolea
+            assertThat(compaction.keyspace())
+                    .as("Active compaction keyspace should not be null")
+                    .isNotNull()
+                    .isNotBlank();
```
`isNotBlank` already covers `isNotNull`. Please keep only the `isNotBlank`

sklaha replied:
Updated.

yifan-c reviewed (commented):
In general, the Cassandra projects prefer `public final` members over `getters` to reduce boilerplate. It would be nice if you can apply the change to your patch, although I am not strongly against what you already have. 
If interested, check out https://cassandra.apache.org/_/development/code_style.html for code style guide details.

sklaha commented:
> In general, the Cassandra projects prefer `public final` members over `getters` to reduce boilerplate. It would be nice if you can apply the change to your patch, although I am not strongly against what you already have. If interested, check out https://cassandra.apache.org/_/development/code_style.html for code style guide details.

Thanks for the link @yifan-c . Most of the classes in `org.apache.cassandra.sidecar.common.response` package have the variables as `private final`. Are you sure that we would like to make the variables in the new classes `public final`?

Commit:
CASSSIDECAR-329: Addressed more review comments

yifan-c commented on server-common/src/main/java/org/apache/cassandra/sidecar/common/server/data/CompactionStatsData.java:
```diff
@@ -0,0 +1,296 @@
+ */
+public class CompactionStatsData
+{
+    private final long concurrentCompactors;
+    private final Map<String, Map<String, Integer>> pendingTasks;
+    private final long totalPendingTasks;
+    private final long completedCompactions;
+    private final long dataCompacted;
+    private final long abortedCompactions;
+    private final long reducedCompactions;
```
Just picking this file for an example. Those members can be changed to `public final` and remove the getters. 

For the response object, the getters are kept merely because of the `@JsonProperty` annotation. 

As mentioned in the last comment, I do not feel strongly against what you already have. I just think that, as a reviewer, I should call it out.

sklaha replied:
I updated the variables in the data objects to be public final.

Commit:
CASSSIDECAR-329:Updated private final fields to public final

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraStorageOperations.java:
```diff
@@ -294,4 +294,24 @@ public void startGossiping()
+    @Override
+    public long getCompactionThroughputBytesPerSec()
+    {
+        LOGGER.warn("getCompactionThroughputBytesPerSec is not supported in Cassandra 4.0");
```
it looks like in 4.0 we have `getCompactionThroughputMbPerSec` . Should we use this and return the value converted to bytes?

sklaha replied:
Thanks for the suggestion. I updated it.

frankgh commented on adapters/adapters-cassandra41/src/main/java/org/apache/cassandra/sidecar/adapters/cassandra41/Cassandra41StorageOperations.java:
```diff
@@ -68,4 +71,14 @@ public void takeSnapshot(@NotNull String tag,
+    public long getCompactionThroughputBytesPerSec()
+    {
+        return jmxClient.proxy(StorageJmxOperations.class, STORAGE_SERVICE_OBJ_NAME)
+                .getCompactionThroughtputBytesPerSec();
```
formatting is off

sklaha replied:
Fixed.

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -61,6 +66,29 @@ public class CassandraMetricsOperations implements MetricsOperations
+    // Comma-separated list of SSTable names involved in the compaction
+    public static final String SSTABLES = "sstables";
+    // Directory where compaction output will be written
+    public static final String TARGET_DIRECTORY = "targetDirectory";
```
`ID` seems unused, also let's move these to `org.apache.cassandra.sidecar.service.CompactionStatsService`

sklaha replied:
Fixed.

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperations.java:
```diff
@@ -250,4 +254,30 @@ private List<ClientConnectionEntry> statsToEntries(Stream<ConnectedClientStats>
+        MeterMetricsJmxOperations metricsProxy = jmxClient.proxy(MeterMetricsJmxOperations.class, metricObjectType);
+
+        return CompletedCompactionsRateData.builder()
+            .meanRate(metricsProxy.getMeanRate())
```
formatting is off here

sklaha replied:
fixed.

frankgh commented on adapters/adapters-base/src/test/java/org/apache/cassandra/sidecar/adapters/base/CassandraMetricsOperationsTest.java:
can you reformat this file with Sidecar's code style?

sklaha replied:
done

frankgh commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/response/data/PendingCompactionTasks.java:
```diff
@@ -0,0 +1,61 @@
+ * Represents pending compaction tasks by keyspace and table
+ */
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class PendingCompactionTasks
```
is this class used at all?

sklaha replied:
removed

frankgh commented on server/src/main/java/org/apache/cassandra/sidecar/service/CompactionStatsService.java:
```diff
@@ -0,0 +1,215 @@
+ * Service responsible for fetching compaction stats from various sources including
+ * CassandraStorageOperations, CassandraMetricsOperations, and CompactionManagerOperations.
+ */
+public class CompactionStatsService
```
this seems misplaced. This should really  live alongside the adapter, because the details are Cassandra-version specific

sklaha replied:
Refactored the code.

frankgh reviewed (commented):
Added some additional comments

Commit:
CASSSIDECAR-329: Refactored code based on review comments

sklaha committed:
Merge branch 'trunk' into CASSSIDECAR-329

Commit:
CASSSIDECAR-329:Fixed integration test failure for 4.0

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/utils/DataTypeUtils.java:
```diff
@@ -0,0 +1,122 @@
+     * @param megabytes the value in megabytes to convert
+     * @return the equivalent value in bytes
+     */
+    public static long megabytesToBytes(long megabytes)
```
the correct terminology is mebibytes, can you please update it?

sklaha replied:
Updated.

frankgh commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/utils/DataTypeUtils.java:
```diff
@@ -0,0 +1,122 @@
+     */
+    public static long megabytesToBytes(long megabytes)
+    {
+        return megabytes * 1024 * 1024;
```
NIT 
```suggestion
        return megabytes << 20;
```

sklaha replied:
Updated.

arjunashok commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraCompactionStatsOperations.java:
```diff
@@ -0,0 +1,233 @@
+                                  .build();
+    }
+
+    private long getCompactionMetricLong(CompactionStatsMetrics metric)
```
Nit: The return type is evident from the method signature and redundant in the method name.

sklaha replied:
Done

arjunashok commented on integration-tests/src/integrationTest/org/apache/cassandra/sidecar/routes/CassandraStatsIntegrationTest.java:
```diff
@@ -256,4 +280,280 @@ void assertClientStatsResponse(HttpResponse<Buffer> response, Map<String, Boolea
+        // Completion rates validation
+        assertThat(stats.completedCompactionsRate()).isNotNull();
+
+        // Validate mean rate format is X.XX/hour
```
Nit: Update comment to reflect the check

sklaha replied:
Fixed

arjunashok commented on integration-tests/src/integrationTest/org/apache/cassandra/sidecar/routes/CassandraStatsIntegrationTest.java:
```diff
@@ -256,4 +280,280 @@ void assertClientStatsResponse(HttpResponse<Buffer> response, Map<String, Boolea
+            }
+        }
+        assertThat(stats).isNotNull();
+        logger.info("Response:{}", stats);
```
Nit: Serialize (`Response:org.apache.cassandra.sidecar.common.response.CompactionStatsResponse@2a9d6a9d`)

sklaha replied:
Added toStrings for the response classes.

arjunashok reviewed (approved):
Few minor nits, but looks good on the whole. +1

Commit:
CASSSIDECAR-329: Addressed nit comments

Commit:
CASSSIDECAR-329: Removed a sys out log

frankgh reviewed (commented):
Looking great. I have some minor comments and I think we should be good with those changes

Commit:
CASSSIDECAR-329: Addressed nit comments

Commit:
CASSSIDECAR-329: Updated mebibytesToBytes method parameter name to mebibytes

frankgh reviewed (approved):
+1
